### DISCUSSION TITLE
Rotary Encoder only increments

### DISCUSSION THREAD

**Discussion by StephanBeutel** [tttapa/Control-Surface#486] [Q&A]
Hi,

I want to build a midi controller for Ableton. Directly after starting the project I stuck, because the rotary encoder only increments the value. I can't decrement. So the turn direction is not mentioned.

I took the code from the example to start with:

```
#include <Encoder.h> // Include the Encoder library.
// This must be done before the Control Surface library.
#include <Control_Surface.h> // Include the Control Surface library
 
// Instantiate a MIDI over USB interface.
//HairlessMIDI_Interface midi;
USBMIDI_Interface midi;

// Instantiate a CCRotaryEncoder object
CCRotaryEncoder enc = {
  {2, 3},       // pins
  MCU::V_POT_1, // MIDI address (CC number + optional channel)
  1,            // optional multiplier if the control isn't fast enough
};
 
void setup() {
  // Select the correct relative MIDI CC mode.
  // Options:
  //   - TWOS_COMPLEMENT (default)
  //   - BINARY_OFFSET
  //   - SIGN_MAGNITUDE
  //   - NEXT_ADDRESS
  // Aliases:
  //   - REAPER_RELATIVE_1
  //   - REAPER_RELATIVE_2
  //   - REAPER_RELATIVE_3
  //   - TRACKTION_RELATIVE
  //   - MACKIE_CONTROL_RELATIVE
  //   - KORG_KONTROL_INC_DEC_1
  RelativeCCSender::setMode(relativeCCmode::TWOS_COMPLEMENT);
  Control_Surface.begin(); // Initialize Control Surface
}
 
void loop() {
  Control_Surface.loop(); // Update the Control Surface
}
```

I also can replicate it when using the USBMIDI_Interface. This is the output from the serial console, 7 outputs for each direction. I inserted a separator in between:

```
Control Change  	Channel: 1	Data 1: 0x10	Data 2: 0x01	Cable: 1
Control Change  	Channel: 1	Data 1: 0x10	Data 2: 0x01	Cable: 1
Control Change  	Channel: 1	Data 1: 0x10	Data 2: 0x01	Cable: 1
Control Change  	Channel: 1	Data 1: 0x10	Data 2: 0x01	Cable: 1
Control Change  	Channel: 1	Data 1: 0x10	Data 2: 0x01	Cable: 1
Control Change  	Channel: 1	Data 1: 0x10	Data 2: 0x01	Cable: 1
Control Change  	Channel: 1	Data 1: 0x10	Data 2: 0x01	Cable: 1
-----
Control Change  	Channel: 1	Data 1: 0x10	Data 2: 0x01	Cable: 1
Control Change  	Channel: 1	Data 1: 0x10	Data 2: 0x01	Cable: 1
Control Change  	Channel: 1	Data 1: 0x10	Data 2: 0x01	Cable: 1
Control Change  	Channel: 1	Data 1: 0x10	Data 2: 0x01	Cable: 1
Control Change  	Channel: 1	Data 1: 0x10	Data 2: 0x01	Cable: 1
Control Change  	Channel: 1	Data 1: 0x10	Data 2: 0x02	Cable: 1
Control Change  	Channel: 1	Data 1: 0x10	Data 2: 0x01	Cable: 1
```

The only difference in each line is sometimes the Data2 value, depending on the turning speed.

Any ideas whats wrong? I tried different rotary encoders, which work fine on other projects.

I use an Arduino Nano 33 IoT and a KY-040 rotary encoder (https://www.az-delivery.de/products/drehimpulsgeber-modul).

**Comment by tttapa** [comment]
This is most likely caused by a hardware or wiring issue.

**Reply by StephanBeutel** [reply]
``My wiring is very simple:

| Nano33  | Rotary Encoder HY-040  |
|---|---|---|---|---|
| Grd  | Grd  | 
| +5V  | + in  |
| D2  | Clk  | 
| D3  | DT  | 

Even if I change D2/D3, result is the same. And +/Grd can't be changed.

I also tried the code with an old Arduino Nano. There I get these outputs for right/left turn:

```
Control Change  	Channel: 1	Data 1: 0x10	Data 2: 0x7f	Cable: 1 (right turn/increment)
Control Change  	Channel: 1	Data 1: 0x10	Data 2: 0x01	Cable: 1 (left turn/decrement)
```

Any ideas how to change wiring?

**Reply by StephanBeutel** [reply]
![image](https://user-images.githubusercontent.com/1589631/119559420-efc5c780-bda2-11eb-925e-c4c207452ed0.jpeg)

I removed the +5V connection, as the functionality is the same without.

**Reply by tttapa** [reply]
The output from the Nano looks alright.

Which version of the library are you using? 1.2.0? `master`? `new-input`?  
Do the examples from the [Encoder library](https://github.com/PaulStoffregen/Encoder) work correctly?

**Reply by StephanBeutel** [reply]
Also the basic example from the encoder library only increments.
Probably my rotary encoder is not compatible. Do you have any suggested rotary encoder which works?

**Reply by StephanBeutel** [reply]
I also tested the basic example with my normal Nano There the increment/decrement is working.

So it must be the Arduino Nano 33 IoT which is the problem.

**Reply by tttapa** [reply]
The Nano 33 IoT uses the same microcontroller as the Arduino Zero, and the Encoder library has the following to say about it:
https://github.com/PaulStoffregen/Encoder/blob/8a3ebcae9f9e99cddb562348809c4be76c7a8c2b/utility/interrupt_pins.h#L253-L254

You could try using the [`new-input`](https://github.com/tttapa/Control-Surface/tree/new-input), it uses a different Encoder library. When you use that version, you have to remove the `#include <Encoder.h>` from your sketch or it will interfere.

**Reply by StephanBeutel** [reply]
Same result with the new-input branch
